File: semester_2/homework_2/Problem1/Stack.cs
namespace StackStuff
{
    using System;

    public class Stack<T> : IStack<T>
    {
        // Size of the stack
        private int size;

        // Reference to the top element of the stack
        private StackElement top;

        // Checks if stack is empty
        public bool IsEmpty() => this.size == 0;

        // Returns top element of the stack (and deletes it)
        public T Pop()
        {
            if (this.IsEmpty())
            {
                throw new InvalidOperationException("Stack is empty!!!");
            }
            
            var returnValue = this.top;

            this.top = this.top.Next;
            --this.size;

            return returnValue.Value;
        }

        // Inserts new element to the stack
        public void Push(T value)
        {
            this.top = new StackElement(value, this.top);
            ++this.size;
        }

        private class StackElement
        {
            public readonly T Value;
            public readonly StackElement Next;

            public StackElement(T value, StackElement next)
            {
                this.Value = value;
                this.Next = next;
            }
        }
    }
}

File: semester_4/hometask_5/Problem1/Stack.fs
namespace Problem1

module Stack = 
    /// Represents simple stack
    type Stack<'a> =
        | EmptyStack
        | StackNode of 'a * Stack<'a>

    /// Checks if stack is empty
    let isEmpty stack =
        match stack with
            | EmptyStack -> true
            | _ -> false

    /// Pushes given value to the given stack
    let push value stack = 
        StackNode(value, stack)

    /// Returns the top value of the given stack
    let top stack =
        match stack with
            | EmptyStack -> failwith "The stack is empty!"
            | StackNode(value, _) -> value

    /// Removes the top value of the given stack
    let pop stack =
        match stack with
            | EmptyStack -> EmptyStack
            | StackNode(value, next) -> next
